src/irmin-pack/pack_key.ml

```diff
@@ -0,0 +1,176 @@
+open! Import
+include Pack_key_intf
+
+module Codec = struct
+  type t = Hash_and_offset | Hash
+end
+
+type 'hash state =
+  | Direct of { hash : 'hash; offset : int63; length : int }
+  | Direct_unknown_length of { hash : 'hash; offset : int63 }
+  | Indexed of 'hash
+
+type 'hash t = { mutable state : 'hash state }
+
+let inspect t = t.state
+
+let to_hash t =
+  match t.state with
+  | Direct t -> t.hash
+  | Direct_unknown_length t -> t.hash
+  | Indexed h -> h
+
+let promote_exn t ~offset ~length =
+  let () =
+    match t.state with
+    | Direct _ ->
+        Fmt.failwith "Attempted to promote a key that is already Direct"
+    | Direct_unknown_length t ->
+        if not (Int63.equal offset t.offset) then
+          Fmt.failwith
+            "Attempted to promote a key with offset %a to one at offset %a"
+            Int63.pp t.offset Int63.pp offset
+    | Indexed _ -> ()
+  in
+  t.state <- Direct { hash = to_hash t; offset; length }
+
+let t : type h. h Irmin.Type.t -> h t Irmin.Type.t =
+ fun hash_t ->
+  let open Irmin.Type in
+  variant "t" (fun direct direct_blindfolded indexed t ->
+      match t.state with
+      | Direct { hash; offset; length } -> direct (hash, offset, length)
+      | Direct_unknown_length { hash; offset } ->
+          direct_blindfolded (hash, offset)
+      | Indexed x1 -> indexed x1)
+  |~ case1 "Direct" [%typ: hash * int63 * int] (fun (hash, offset, length) ->
+         { state = Direct { hash; offset; length } })
+  |~ case1 "Direct_unknown_length" [%typ: hash * int63] (fun (hash, offset) ->
+         { state = Direct_unknown_length { hash; offset } })
+  |~ case1 "Indexed" [%typ: hash] (fun x1 -> { state = Indexed x1 })
+  |> sealv
+
+let t (type hash) (codec : Codec.t) (hash_t : hash Irmin.Type.t) =
+  let module Hash = struct
+    type t = hash
+    [@@deriving irmin ~equal ~compare ~pre_hash ~encode_bin ~decode_bin]
+
+    let encoded_size =
+      match Irmin.Type.Size.of_value t with
+      | Static n -> n
+      | Dynamic _ | Unknown ->
+          Fmt.failwith "Hash must have a fixed-width binary encoding"
+  end in
+  (* The pre-hash image of a key is just the hash of the corresponding value.
+
+     NOTE: it's particularly important that we discard the file offset when
+     computing hashes of structured values (e.g. inodes), so that this hashing
+     process is reproducible in different stores (w/ different offsets for the
+     values). *)
+  let equal a b =
+    let result = Hash.equal (to_hash a) (to_hash b) in
+    assert (
+      if not result then true
+      else
+        match (a.state, b.state) with
+        | Direct a, Direct b -> a.length = b.length
+        | _, _ -> true);
+    result
```

Comment:
> If a piece of code ever tries to test the equality of 2 keys with the same hash but with different offsets, the false result sounds more predictable to me than true.

I don't feel this way because I consider `Repr.{equal,compare}` as providing some notion of (cache-agnostic) _structural_ equality – as opposed to physical equality – so that the equalities compose meaningfully into larger types. Two `Direct` keys with the same hash but different offsets should be considered equal (as in [this test case](https://github.com/CraigFe/irmin/blob/irmin-pack-keys-with-offsets/test/irmin/generic-key/test_store_offset.ml#L42-L43)), as they're references to the same (albeit duplicated) object.

e.g. if I derive equality for `Node.t`, I would expect `equal_node n1 n2` to hold if `n1` and `n2` have the same structure / hash, and not be sensitive to whether the child pointers reference the same instances of a duplicated object. Similarly, I'd want `equal_node` to not depend on whether the internal keys are `Direct` or `Indexed`, as this is an implementation detail. (This is the way that `Repr.equal Tree.t` is supposed to behave too.)

Finally, falling back to hash-based equality on `equal (Indexed _) (Direct _)` feels wrong to me: perhaps the indexed object is not at the same offset as the `Direct` one, so the two keys would be considered non-equal if the user had previously dereferenced both keys.

Regardless of what semantics we want, we should document it [here](https://github.com/CraigFe/irmin/blob/irmin-pack-keys-with-offsets/src/irmin/key.mli#L86-L89), where there's already some discussion of `pre_hash key_t`.

Reply:
I understand your point. Documenting it should be enough

Thanks you for taking time to answer with precision